File: Course 1 - Algorithm Toolbox/week3_greedy_algorithms/6_maximum_number_of_prizes/different_summands.py
# Uses python3
import sys

def optimal_summands(num_provided):
    num = 0
    numbers_picked = set()
    while num_provided != 0:
        num += 1
        num_provided -= num
        numbers_picked.add(num)
        if num_provided in numbers_picked:
            numbers_picked.remove(num)
            num_provided += num
            numbers_picked.add(num_provided)
            break
    return numbers_picked

if __name__ == '__main__':
    input = sys.stdin.read()
    n = int(input)
    summands = optimal_summands(n)
    print(len(summands))
    for x in summands:
        print(x, end=' ')

File: Course 1 - Algorithm Toolbox/week3_greedy_algorithms/6_maximum_number_of_prizes/different_summands.cpp
#include <iostream>
#include <vector>

using std::vector;

vector<int> optimal_summands(int n) {
  int num = 0;
  vector<int> summands;
  while (n != 0)
  {
      num += 1;
      n = n - num;
      summands.push_back(num);
      if (std::count(summands.begin(), summands.end(), n) != 0)
      {
          summands.pop_back();
          n = n + num;
      }
  }
  return summands;
}

int main() {
  int n;
  std::cin >> n;
  vector<int> summands = optimal_summands(n);
  std::cout << summands.size() << '\n';
  for (size_t i = 0; i < summands.size(); ++i) {
    std::cout << summands[i] << ' ';
  }
}
